Split into CLI and LIB packages, use soname in lib package version

sdvillal opened the pull request:
Alternative to #18.

@183amir @isuruf I would be happy if you can review.

- Is this what you have in mind @isuruf?
- I do not like how we need to hardcode the file paths. Any alternative?
- I do not know if I got right run_exports.
- I think downstream dependencies most likely want to depend on libx264 (that is the logical thing to do). I will add a second commit trying to go that path.

I will try to carry the same tests I did for #18, namely that pins are respected downstream.

sdvillal committed:
libx264 vs h264 split, use soname in lib version

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found some lint.

Here's what I've got...


For **recipe**:

 * The recipe must have some tests.

sdvillal commented:
mmm the linter is wrong here? we do have tests on each output...

sdvillal commented:
Linter bug: https://github.com/conda-forge/conda-smithy/issues/833

Will workaround it here.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found some lint.

Here's what I've got...


For **recipe**:

 * The top level meta keys are in an unexpected order. Expecting ['package', 'source', 'build', 'requirements', 'test', 'outputs', 'about', 'extra'].

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

sdvillal commented:
Travis mac builds seem fundamentally broken (we are not the only PR with this problem).

sdvillal commented:
Travis error: https://github.com/conda-forge/conda-smithy/issues/834

sdvillal commented:
I have put run_exports in libx264 (as downstream only need to depend on libx264 and not h264), rebased, uploaded packages (these and ffmpeg) to my channel (sdvillal), where also live packages from #18, and run some manual tests. From my side, this is ready to go after we fix anything you find.

I do not mind either #18 or #19. I like more #18 because I feel it is simpler and it fully avoids having two versions of x264 installed concurrently. Example:

```sh
# This will both install an old version of x264 and the required version of libx264
# The x264 executable will be the old one
# Will have different so versions
# Will have headers clobbed
conda create -c sdvillal -n h264-mess ffmpeg x264=20131217

# The same problem will happen if updating ffmpeg on an environment that already has
# an old version of h264
```

183amir commented:
I prefer #18 too since that's what other distros are doing and as you said there are no clobber issues.

mariusvniekerk commented on recipe/meta.yaml:
```diff
@@ -25,13 +23,40 @@ requirements:
+    version: {{ lib_version }}
+    build:
+      run_exports:
+        - libx264 >={{ so_number }}.00000000,<{{ so_number|int + 1 }}.00000000
```
shouldn't we use `release_date` here?

sdvillal replied:
the reason we don't is that release date has little to do with version compatibility

183amir replied:
You should use `{{ pin_subpackage('libx264') }}` here same as above. but I really prefer we don't do this and go ahead with #18 instead.

sdvillal committed:
use pin_subpackage

the trouble with this is that the generated pin lower-bound will depend also on date,
which is not a big problem but it is not needed either

I guess we could also change min_pin, but nothing it is not the trouble

sdvillal commented:
We have merged #18
